x/staking/client/rest/grpc_query_test.go

```diff
@@ -96,6 +119,846 @@ func (s *IntegrationTestSuite) TestQueryValidatorsGRPCHandler() {
+	delAddressBase64 := base64.URLEncoding.EncodeToString(val.Address)
+
+	testCases := []struct {
+		name       string
+		url        string
+		headers    map[string]string
+		error      bool
+		ubdsLength int
+	}{
+		{
+			"wrong validator address",
+			fmt.Sprintf("%s/cosmos/staking/v1beta1/delegators/%s/unbonding_delegations", baseURL, "wrongValAddress"),
+			map[string]string{
+				grpctypes.GRPCBlockHeightHeader: "2",
+			},
+			true,
+			0,
+		},
+		{
+			"with no validator address",
+			fmt.Sprintf("%s/cosmos/staking/v1beta1/delegators/%s/unbonding_delegations", baseURL, ""),
+			map[string]string{
+				grpctypes.GRPCBlockHeightHeader: "2",
+			},
+			true,
+			0,
+		},
+		{
+			"valid request with less height",
+			fmt.Sprintf("%s/cosmos/staking/v1beta1/delegators/%s/unbonding_delegations", baseURL, delAddressBase64),
+			map[string]string{
+				grpctypes.GRPCBlockHeightHeader: "2",
+			},
+			false,
+			0,
+		},
+		{
+			"valid request",
+			fmt.Sprintf("%s/cosmos/staking/v1beta1/delegators/%s/unbonding_delegations", baseURL, delAddressBase64),
+			map[string]string{
+				grpctypes.GRPCBlockHeightHeader: "3",
+			},
+			false,
+			1,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		s.Run(tc.name, func() {
+			resp, err := testutil.GetRequestWithHeaders(tc.url, tc.headers)
+			s.Require().NoError(err)
+
+			var ubds types.QueryDelegatorUnbondingDelegationsResponse
+
+			err = val.ClientCtx.JSONMarshaler.UnmarshalJSON(resp, &ubds)
+
+			if tc.error {
+				s.Require().Error(err)
+			} else {
+				s.Require().NoError(err)
+				s.Require().Len(ubds.UnbondingResponses, tc.ubdsLength)
+			}
+		})
+	}
+}
+
+func (s *IntegrationTestSuite) TestQueryRedelegationsGRPC() {
+	val := s.network.Validators[0]
+	val2 := s.network.Validators[1]
+	baseURL := val.APIAddress
+
+	// TODO: need to pass bech32 string instead of base64 encoding string
+	delAddressBase64 := base64.URLEncoding.EncodeToString(val.Address)
+
+	// TODO: need to pass bech32 string instead of base64 encoding string
+	srcValBase64 := base64.URLEncoding.EncodeToString(val.Address)
+
+	// TODO: need to pass bech32 string instead of base64 encoding string
+	dstValBase64 := base64.URLEncoding.EncodeToString(val2.Address)
+
+	testCases := []struct {
+		name    string
+		url     string
+		headers map[string]string
+		error   bool
+	}{
+		{
+			"wrong validator address",
+			fmt.Sprintf("%s/cosmos/staking/v1beta1/delegators/%s/redelegations", baseURL, "wrongValAddress"),
+			map[string]string{
+				grpctypes.GRPCBlockHeightHeader: "2",
+			},
+			true,
+		},
+		{
+			"with no validator address",
+			fmt.Sprintf("%s/cosmos/staking/v1beta1/delegators/%s/redelegations", baseURL, ""),
+			map[string]string{
+				grpctypes.GRPCBlockHeightHeader: "2",
+			},
+			true,
+		},
+		{
+			"valid request",
+			fmt.Sprintf("%s/cosmos/staking/v1beta1/delegators/%s/redelegations", baseURL, delAddressBase64),
+			map[string]string{
+				grpctypes.GRPCBlockHeightHeader: "3",
```

Comment:
I see here you are using 2 and 3 as header height. Is there some different result expected at height 2 vs 3?

Reply:
I'm cleaning some stuff, thanks for pointing.